File: epi_judge_python/replace_and_remove.py
import functools

from test_framework import generic_test
from test_framework.test_utils import enable_executor_hook


def replace_and_remove(size, s):
    clean = 0
    num_a = 0
    for i in range(size):
        x = s[i]
        if x == "a":
            num_a += 1
        if x != "b":
            s[clean] = x
            clean += 1

    res = right_end = num_a + clean - 1
    for i in reversed(range(clean)):
        x = s[i]
        if x == "a":
            s[right_end], s[right_end - 1] = "d", "d"
            right_end -= 2
        else:
            s[right_end] = x
            right_end -= 1

    return res + 1

@enable_executor_hook
def replace_and_remove_wrapper(executor, size, s):
    res_size = executor.run(functools.partial(replace_and_remove, size, s))
    return s[:res_size]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("replace_and_remove.py",
                                       'replace_and_remove.tsv',
                                       replace_and_remove_wrapper))

File: epi_judge_java/epi/ReplaceAndRemove.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TimedExecutor;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
public class ReplaceAndRemove {

  public static int replaceAndRemove(int size, char[] s) {
    int relevant = 0, aCount = 0;

    for (int i = 0; i < size; i++){

      if (s[i] != 'b'){
        if (s[i] == 'a') aCount++;
        char tmp = s[relevant];
        s[relevant++] = s[i];
        s[i] = tmp;
      }
    }
    relevant--;
    int r = relevant + aCount;
    int fullSize = r;
    while (relevant >= 0){
      if(s[relevant] == 'a'){
        s[r--] = 'd';
        s[r--] = 'd';
      } else {
        s[r--] = s[relevant];
      }
      relevant--;
    }

    return fullSize + 1;
  }
  @EpiTest(testDataFile = "replace_and_remove.tsv")
  public static List<String>
  replaceAndRemoveWrapper(TimedExecutor executor, Integer size, List<String> s)
      throws Exception {
    char[] sCopy = new char[s.size()];
    for (int i = 0; i < size; ++i) {
      if (!s.get(i).isEmpty()) {
        sCopy[i] = s.get(i).charAt(0);
      }
    }

    Integer resSize = executor.run(() -> replaceAndRemove(size, sCopy));

    List<String> result = new ArrayList<>();
    for (int i = 0; i < resSize; ++i) {
      result.add(Character.toString(sCopy[i]));
    }
    return result;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "ReplaceAndRemove.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
